include/boost/geometry/algorithms/detail/overlay/get_clusters.hpp

```diff
@@ -0,0 +1,302 @@
+
+    // Compare points with each other. Performance is O(n log(n)) because of the sorting.
+    int cnt = 0;
+    for (auto it1 = points.begin(); it1 != points.end(); ++it1)
+    {
+        // Inner loop, walks until too far away in y-direction
+        for (auto it2 = it1 + 1; it2 != points.end(); ++it2)
+        {
+            auto const d = geometry::get<1>(it1->pnt) - geometry::get<1>(it2->pnt);
+            if (equal_policy.exceeds(d))
+            {
+                // Too far away in y-direction to be considered as equal
+                break;
+            }
+            cnt++;
+            if (equal_policy.equals(it1->pnt, it2->pnt))
+            {
+                std::size_t cindex = 0;
+                // Most recent clusters (with this y-value) are at the bottom
+                // therefore we can stop as soon as the y-value is out of reach (TODO)
+                bool found = false;
+                for (auto cit = clustered_points.begin(); cit != clustered_points.end(); ++cit)
+                {
+                    found = equal_policy.equals(cit->pnt, it1->pnt);
+                    if (found)
+                    {
+                        break;
+                    }
+                    cindex++;
+                }
+
+                // Add new cluster
+                if (! found)
+                {
+                   cindex = clustered_points.size();
+                   cluster_type newcluster;
+                   clustered_points.push_back({newcluster, it1->pnt});
+                }
+                clustered_points[cindex].cluster.turn_indices.insert(it1->turn_index);
+                clustered_points[cindex].cluster.turn_indices.insert(it2->turn_index);
+            }
+        }
+    }
+
+    // Convert to map
+    signed_size_type cluster_id = 1;
+    for (auto& trace : clustered_points)
+    {
+        clusters[cluster_id++] = trace.cluster;
+    }
+}
+
+template
+<
+    typename Turns,
+    typename Clusters
+>
+inline void assign_cluster_ids(Turns& turns, Clusters const& clusters)
+{
+    for (auto& turn : turns)
+    {
+        turn.cluster_id = -1;
+    }
+    for (auto const& kv : clusters)
+    {
+        for (const auto& index : kv.second.turn_indices)
+        {
+            turns[index].cluster_id = kv.first;
+        }
+    }
+}
+
+template <typename Geometry0, typename Geometry1>
+inline bool is_consecutive(segment_identifier const& first, segment_identifier const& second,
+                           Geometry0 const& geometry0, Geometry1 const& geometry1)
+{
+    if (first.source_index == second.source_index
+        && first.ring_index == second.ring_index
+        && first.piece_index == second.piece_index
+        && first.multi_index == second.multi_index)
+    {
+        auto const sd = first.source_index == 0
+                        ? segment_distance(geometry0, first, second)
+                        : segment_distance(geometry1, first, second);
+        return sd == 1;
```

Comment:
This code looks confusing to me. The function naming suggests that some kind distance is calculated. Since the geometry is passed and `segment` is in the name I jump to the conclusion that this is some kind of geometric distance. We can't deduce what's happening from the return type neither because `auto` is used. On the other hand the distance is compared with `1` which makes little sense.

I think that explicitly stating the type `signed_size_type` would be a good start. I guess this name refers to the `std::distance()` but if we could change the name to something more descriptive I think this would be beneficial.

Reply:
It's the distance in terms of segments (including circular behavior). So two segments are consecutive if that distance is 1.
I actually don't see much confusion here. Of course I can replace auto with `signed_size_type` indeed.

Reply:
:heavy_check_mark: